pints/_mcmc/_nuts.py

```diff
@@ -0,0 +1,557 @@
+#
+# No-U-Turn Sampler (NUTS) MCMC method
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import asyncio
+import pints
+import numpy as np
+
+
+class NutsState:
+    """
+    Class to hold information about the current state of the NUTS hamiltonian
+    integration path
+
+    NUTS builds up the integration path implicitly via recursion up a binary tree, this
+    class handles combining states from different subtrees (see `update`)
+
+    NUTS integrates both backwards ("minus") and forwards ("plus"), so this state must
+    keep track of both end points of the integration path
+
+    Attributes
+    ----------
+
+    theta_minus: ndarray
+        parameter value at the backwards end of the integration path
+
+    theta_plus: ndarray
+        parameter value at the forwards end of the integration path
+
+    r_minus: ndarray
+        momentum value at the backwards end of the integration path
+
+    r_plus: ndarray
+        momentum value at the forwards end of the integration path
+
+    L_minus: float
+        logpdf value at the backwards end of the integration path
+
+    L_plus: float
+        logpdf value at the forwards end of the integration path
+
+    grad_L_minus: float
+        gradient of logpdf at the backwards end of the integration path
+
+    grad_L_plus: float
+        gradient of logpdf at the forwards end of the integration path
+
+    n: int
+        number of accepted points in the path
+
+    s: int
+        0 if sufficient leapfrog steps have been taken, 1 otherwise
+
+    theta: ndarray
+        the current accepted point along the path
+
+    L: float
+        the logpdf of the current accepted point
+
+    grad_L: float
+        the gradient of the logpdf at the current accepted point
+
+    alpha: float
+        the acceptance probability
+
+    n_alpha: float
+        a count of the points along this path
+
+    """
+
+    def __init__(self, theta, r, L, grad_L, n, s, alpha, n_alpha):
+        self.theta_minus = theta
+        self.theta_plus = theta
+        self.r_minus = r
+        self.r_plus = r
+        self.L_minus = L
+        self.L_plus = L
+        self.grad_L_minus = grad_L
+        self.grad_L_plus = grad_L
+        self.n = n
+        self.s = s
+        self.theta = theta
+        self.L = L
+        self.grad_L = grad_L
+        self.alpha = alpha
+        self.n_alpha = n_alpha
+
+    def update(self, other_state, direction, root):
+        """
+        if root == True, this combines a depth j subtree (`self`) with a depth j+1
```

Comment:
` ``root == True`` ` etc (double ticks in RST)

Reply:
done